actors/market/src/state.rs

```diff
@@ -551,20 +572,163 @@ impl State {
+    }
+
+    // Reads and removes the sector deals mapping for an array of sector numbers,
+    pub fn pop_sector_deal_ids<BS>(
+        &mut self,
+        store: &BS,
+        provider: &Address,
+        sector_numbers: impl Iterator<Item = SectorNumber>,
+    ) -> Result<Vec<(SectorNumber, SectorDealIDs)>, ActorError>
+    where
+        BS: Blockstore,
+    {
+        let mut provider_sectors = self.load_provider_sectors(store)?;
+        let mut sector_deals = load_provider_sector_deals(store, &provider_sectors, provider)?;
+
+        let mut popped_sector_deals = Vec::new();
+        for sector_number in sector_numbers {
+            let deals: Option<SectorDealIDs> = sector_deals
+                .delete(&sector_number)
+                .with_context(|| format!("provider {}", provider))?;
+            if let Some(deals) = deals {
+                popped_sector_deals.push((sector_number, deals.clone()));
+            }
+        }
+
+        // Flush if any of the requested sectors were found.
+        if !popped_sector_deals.is_empty() {
+            if sector_deals.is_empty() {
+                // Remove from top-level map
+                provider_sectors
+                    .delete(provider)
+                    .with_context_code(ExitCode::USR_ILLEGAL_STATE, || {
+                        format!("failed to delete sector deals for {}", provider)
+                    })?;
+            } else {
+                save_provider_sector_deals(&mut provider_sectors, provider, &mut sector_deals)?;
+            }
+            self.save_provider_sectors(&mut provider_sectors)?;
+        }
+
+        Ok(popped_sector_deals)
+    }
 
-        Ok(rval_pending_deal)
+    // Removes specified deals from the sector deals mapping.
+    // Missing deals are ignored.
+    pub fn remove_sector_deal_ids<BS>(
+        &mut self,
+        store: &BS,
+        provider_sector_deal_ids: &BTreeMap<Address, BTreeMap<SectorNumber, Vec<DealID>>>,
+    ) -> Result<(), ActorError>
+    where
+        BS: Blockstore,
+    {
+        let mut provider_sectors = self.load_provider_sectors(store)?;
+        for (provider, sector_deal_ids) in provider_sector_deal_ids {
+            let mut sector_deals = load_provider_sector_deals(store, &provider_sectors, provider)?;
+            for (sector_number, deals_to_remove) in sector_deal_ids {
+                let existing_deal_ids = sector_deals
+                    .get(sector_number)
+                    .context_code(ExitCode::USR_ILLEGAL_STATE, "failed to read sector deals")?;
+                if let Some(existing_deal_ids) = existing_deal_ids {
+                    // The filter below is a linear scan of deals_to_remove.
+                    // This is expected to be a small list, often a singleton, so is usually
+                    // pretty fast.
+                    // Loading into a HashSet could be an improvement for large collections of deals
+                    // in a single sector being removed at one time.
```

Comment:
Won't this always be all the deals in the sector? Given that:

1. Cron is paying for it.
2. An SP can create a sector with a _lot_ of deals (thousands?).

This makes be a bit uncomfortable.

If we don't expect missing deals, we can sort then filter:

```rust
                    deals_to_remove.sort();
                    let deals_to_remove = deals_to_remove.iter().peekable();
                    let new_deals = existing_deal_ids // assume sorted?
                        .deals
                        .iter()
                        .filter(|deal_id| deals_to_remove.next_if_eq(deal_id).is_none())
                        .cloned()
                        .collect();
```

(we _can_ skip missing deals, it's just slightly less clean).

Reply:
> Won't this always be all the deals in the sector?
Only if they all expire at around the same epoch. 

Also there is a policy limit of 256 deals / sector, at least in the current precommit-based and non-DDO flow. You've inadvertently pointed to an interesting item that there is _not_ a policy limit on the number of piece manifests activated with an individual sector in the new flow. This is relevant to concurrent discussion about removing artificial policy limits that could be better limited by gas.

With FIP-0074 cron will no longer pay for this (for new deals). The deal _client_ will specify the deals to settle, which becomes deals_to_remove. We need to non-trivially rebase that code on FIP-0076 after landing this, but I expect the issue to become pretty minor.

Reply:
> Only if they all expire at around the same epoch.

It's more a matter of it being under the control of the miner (who can self-deal here).

> Also there is a policy limit of 256 deals / sector

Ok, that makes me feel a lot better (although `256^2` is still a fair amount).

> With FIP-0074 cron will no longer pay for this (for new deals). The deal client will specify the deals to settle, which becomes deals_to_remove. We need to non-trivially rebase that code on FIP-0076 after landing this, but I expect the issue to become pretty minor.

Ah, is that slated for this upgrade? If so, I have no more concerns here.